build/configs/common_download.sh

```diff
@@ -0,0 +1,440 @@
+#!/usr/bin/env bash
+###########################################################################
+#
+# Copyright 2021 Samsung Electronics All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+# common_download.sh
+
+CURDIR=$(readlink -f "$0")
+CUR_PATH=$(dirname "$CURDIR")
+
+#When the location of this script is changed, only TOP_PATH needs to be changed.
+TOP_PATH=${CUR_PATH}/../.. 
+OS_PATH=${TOP_PATH}/os
+BIN_PATH=${TOP_PATH}/build/output/bin
+
+CONFIG=${OS_PATH}/.config
+source ${CONFIG}
+SMARTFS_BIN_PATH=${BIN_PATH}/${CONFIG_ARCH_BOARD}_smartfs.bin
+
+BOARD_CONFIG=${TOP_PATH}/build/configs/${CONFIG_ARCH_BOARD}/board_metadata.txt
+BOARD_SPECIFIC_SCRIPT=${TOP_PATH}/build/configs/${CONFIG_ARCH_BOARD}/${CONFIG_ARCH_BOARD}_download.sh
+
+source ${BOARD_SPECIFIC_SCRIPT}
+source ${BOARD_CONFIG}
+
+## For manual port selection, set $1 as port, set $2 as the portname, eg. make download port ttyUSB2 ALL
+
+if [ "$1" == "port" ]; then
+	PORT="$2"
+	shift 2
+else
+	PORT=${DEFAULT_PORT}
+fi
```

Comment:
abhishek@abhishek-VirtualBox:~/projects/Forked/TizenRT/os$ sudo make download -x ttyUSB1 all
make: invalid option -- 'x'
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.

Giving an option like -p is triggering make options. So I think that "sudo make download -p ttyUSB1 all" is not feasible.

Reply:
Ok, then how about giving `port=xxxx`? What I want is not fixed location for port setting.
I mean
```
$ make download all port=ttyUSB1
or
$ make download port=ttyUSB1 all
```
both should work.

Reply:
Current code works only
```
$ make download port ttyUSB1 all
```
Below does not work
```
$ make download all port ttyUSB1
```

Reply:
Done. Now commands like these are acceptable.

>$ make download all port=ttyUSB1
or
$ make download port=ttyUSB1 all

Reply:
It's not applied. 
Please check what I said again.

Reply:
I have made the changes in Makefile.unix. Since we are getting the arguments from the Makefile.unix I checked if we are passing port=value and then passed it as the first argument and the value as the second argument to the common download script.

changes in os/Makefile.unix :
 

> checking if the port is passed in args like "make download port=ttyusb1" all or "make download all port=ttyusb1" 

	$(Q) if [ -z ${port} ];then  
		$(call DOWNLOAD, $(DOWNLOAD_ARGS)) ; 
	else 

> if the port is passed set $1 to port and $2 to its value.

		$(call DOWNLOAD, port ${port} $(DOWNLOAD_ARGS)) ; 
	fi

Reply:
Oh yeah. You're right. Then could you add that content in download help message? 
Because there is no explanation of port, it is much difficult to know.
```
##Help utility##
function dwld_help()
{
	cat <<EOF
	HELP:
		make download ALL or [PARTITION(S)]
		make download ERASE ALL or erase [PARTITION(S)]
	PARTITION(S):
		[${uniq_parts[@]}]  NOTE:case sensitive
	For examples:
		make download ALL
		make download ERASE ALL
		make download kernel
		make download erase kernel
		make download ota
		make download erase ota
		make download kernel userfs
		make download erase userfs
EOF
}
```

Reply:
I mean adding how to set port in help message.

Reply:
Done. Added these commands  in help :

```
make download ALL port=ttyUSB1
make download port=ttyUSB1 ALL
make download erase port=ttyUSB1 userfs
```

Reply:
```
make download erase port=ttyUSB1 userfs
```
Is this working? Doesn't `port=ttyUSB1' be ` $(DOWNLOAD_ARGS)`?
I mean above becomes like below?
```
make download port ttyUSB1 erase port=ttyUSB1 userfs
```

Reply:
Before passing the download args, they are calling eval on the args. So args like port=ttyUSB1 get evaluated to nothing. Hence such args are not being passed.
```
ifeq (download,$(firstword $(MAKECMDGOALS)))
	# use the rest as arguments for "download"
	DOWNLOAD_ARGS := $(wordlist 2,$(words $(MAKECMDGOALS)),$(MAKECMDGOALS))
 	# Turn DOWNLOAD_ARGS into do-nothing targets
	$(eval $(DOWNLOAD_ARGS):;@:)
endif

download:
```